### PR TITLE
Use `pre-commit` 

### PR DESCRIPTION
**Pull request by pat-s**
https://pre-commit.com

- Applies fixes/lints during local development (if installed locally)
- CI: applies autofixes to PRs using the hooks and config in place (see https://pre-commit.ci/)

- [x] Updated `CONTRIBUTING.md`
- [x] Added README badge
- [x] Added `markdownlint.yaml` config

CI needs to be enabled at https://pre-commit.ci/ by an owner.
See the pre-commit config for initial hooks which would run. NB: they fail currently, as a lot of files are not compliant and require fixes.

Overall this should improve the code style and prevent "cleanup" commits which aim to align/fix files occasionally.

### CODE REVIEW

File: .pre-commit-config.yaml
Comment (pat-s) [reply .pre-commit-config.yaml:24]: Needs to be installed locally or it will complain that it's not found. Overall users are forced to install these tools themselves, you can view them as "developer dependencies". 

If not (i.e. local hooks are not executed), precommit.ci should still do it/fail.

File: CONTRIBUTING.md
Comment (6543) [inline CONTRIBUTING.md:24]: that duble empty spaces at the end are intentional newlines!!!

Comment (pat-s) [reply CONTRIBUTING.md:24]: I know. But IMO one of the worst inventions ever done given the cleanup of whitespaces in almost any other language.
Linebreaks also do the job, i.e. one sentence per line.
Such trailing whitespaces are often not even shown in editors unless the config is explicitly defined to do so.

How about we just clean up all MD docs and go with "one sentence per line"?

Comment (pat-s) [reply CONTRIBUTING.md:24]: There's also the option to append a `\` at the end which is not affected by this.

Comment (xoxys) [reply CONTRIBUTING.md:24]: I would vote for `\` as its visible regardless of editor settings.

File: .pre-commit-config.yaml
Comment (xoxys) [reply .pre-commit-config.yaml:25]: This one is using `language: golang`, so go needs to be installed but not `checkmake` itself, `pre-commit` will prepare this environment automatically, see https://pre-commit.com/#golang

File: .pre-commit-config.yaml
Comment (xoxys) [reply .pre-commit-config.yaml:29]: This one will fail as `hadolint` is using `language: system` in the hook config https://github.com/hadolint/hadolint/blob/master/.pre-commit-hooks.yaml#L11 

`system` requires dependencies to be installed locally already, see https://pre-commit.com/#system

**Comment by pat-s** [comment]
So seems like so far everybody is rather in favor of it and nobody against?
We can always tweak/adapt the config.

@anbraten An owner would need to opt-in to pre-commit.ci if this should also be used.

If runs on their own CI resources and should be quicker than any WP CI check - and fail if people didn't use pre-commit locally before.

File: .pre-commit-config.yaml
Comment (xoxys) [reply .pre-commit-config.yaml:29]: Absolutely not, but we could use the container version instead to make it work out of the box most of the time, https://github.com/hadolint/hadolint/blob/master/.pre-commit-hooks.yaml#L2-L7

**Comment by xoxys** [comment]
Do we want to run markdownlint once across the entire repo to ensure it? Otherwise, the pre-commit hook will fail "randomly" when people touch markdown files (works for me as well).

File: .pre-commit-config.yaml
Comment (pat-s) [reply .pre-commit-config.yaml:29]: I thought about this but wouldn't this mean that docker must be running to use it? I usually have it off by default as it takes a lot of battery and would actually prefer using the system executable instead.

Comment (pat-s) [reply .pre-commit-config.yaml:29]: I guess it depends on who you ask - everyone might have a different preference here 🙃

**Comment by pat-s** [comment]
There are a bunch of files which require updates if we would use this, yes. I left them out on purpose in this PR so the diff is focused on the config files.

After this got merged, we should update all files in a PR so they are hook compliant.

**Review by xoxys — APPROVED** [review]

**Comment by pat-s** [comment]
Cool. I'll merge tomorrow, after I have the PR with the required file changes ready,

**Pull request merged**
